src/Database/Oracle/Simple/Queue.hs

```diff
@@ -0,0 +1,326 @@
+    -- | dpiQueue *
+    DPIQueue ->
+    -- | numProps *
+    Ptr CUInt ->
+    -- | props **
+    Ptr DPIMsgProps ->
+    IO CInt
+
+enqOne :: DPIQueue -> DPIMsgProps -> IO ()
+enqOne dpiQueue dpiMsgProps = 
+    throwOracleError =<<
+        dpiQueue_enqOne dpiQueue dpiMsgProps
+
+foreign import ccall unsafe "dpiQueue_enqOne"
+  dpiQueue_enqOne ::
+    -- | dpiQueue *
+    DPIQueue ->
+    -- | props *
+    DPIMsgProps ->
+    IO CInt
+
+getDeqOptions :: DPIQueue -> IO DPIDeqOptions
+getDeqOptions dpiQueue = do
+  alloca $ \dpiDeqOptionsPtr -> do
+    throwOracleError =<< dpiQueue_getDeqOptions dpiQueue dpiDeqOptionsPtr
+    peek dpiDeqOptionsPtr
+
+foreign import ccall unsafe "dpiQueue_getDeqOptions"
+  dpiQueue_getDeqOptions ::
+    -- | dpiQueue *
+    DPIQueue ->
+    -- | options **
+    Ptr DPIDeqOptions ->
+    IO CInt
+
+getEnqOptions :: DPIQueue -> IO DPIEnqOptions
+getEnqOptions dpiQueue = do 
+  alloca $ \dpiEnqOptionsPtr -> do
+    throwOracleError =<< dpiQueue_getEnqOptions dpiQueue dpiEnqOptionsPtr
+    peek dpiEnqOptionsPtr
+
+foreign import ccall unsafe "dpiQueue_getEnqOptions"
+  dpiQueue_getEnqOptions ::
+    -- | dpiQueue *
+    DPIQueue ->
+    -- | options **
+    Ptr DPIEnqOptions ->
+    IO CInt
+
+queueRelease :: DPIQueue -> IO ()
+queueRelease dpiQueue = throwOracleError =<< dpiQueue_release dpiQueue
+
+foreign import ccall unsafe "dpiQueue_release"
+  dpiQueue_release ::
+    -- | dpiQueue *
+    DPIQueue ->
+    IO CInt
+
+genJSONQueue :: Connection -> String -> IO DPIQueue
+genJSONQueue (Connection fptr) queueName = do
+  withForeignPtr fptr $ \conn -> do
+    alloca $ \dpiQueuePtr -> do
+      withCStringLen queueName $ \(queueNameC , fromIntegral -> queueNameLen) -> do
+        throwOracleError =<< dpiConn_newJsonQueue conn queueNameC queueNameLen dpiQueuePtr
+        peek dpiQueuePtr
+
+foreign import ccall unsafe "dpiConn_newJsonQueue"
+  dpiConn_newJsonQueue ::
+    -- | dpiConn *
+    Ptr DPIConn ->
+    -- | char* name 
+    CString ->
+    -- | name Length
+    CUInt ->
+    -- | dpiQueue **
+    Ptr DPIQueue ->
+    IO CInt
+
+genMsgProps :: Connection -> IO DPIMsgProps
+genMsgProps (Connection fptr) = do
+  withForeignPtr fptr $ \conn -> do
+    alloca $ \dpiMsgPropsPtr -> do
+      throwOracleError =<< dpiConn_newMsgProps conn dpiMsgPropsPtr
+      peek dpiMsgPropsPtr
+
+foreign import ccall unsafe "dpiConn_newMsgProps"
+  dpiConn_newMsgProps ::
+    -- | dpiConn *
+    Ptr DPIConn ->
+    -- | dpiMsgProps **
+    Ptr DPIMsgProps ->
+    IO CInt
+
+-- For now, Passing objectType will not work :( use setPayLoadType.
+genQueue :: Connection -> String -> IO DPIQueue
+genQueue (Connection fptr) queueName = do
+  withForeignPtr fptr $ \conn -> do
+    alloca $ \dpiQueuePtr -> do
+      withCStringLen queueName $ \(queueNameC , fromIntegral -> queueNameLen) -> do
+        throwOracleError =<< dpiConn_newQueue conn queueNameC queueNameLen nullPtr dpiQueuePtr
+        -- TODO: Accomodate ObjectType
```

Comment:
Just curious why is `ObjectType` not supported?

Reply:
I'm Just trying to find a cleaner way of using ObjectType. I will add it soon.